recirq/qml_lfe/learn_dynamics_C.py

```diff
@@ -0,0 +1,167 @@
+flags.DEFINE_integer("depth", None, "Circuit depth (block-wise).")
+flags.DEFINE_integer(
+    "n_data",
+    20,
+    "Number of circuits generated for each class (total circuits == 2 * n_data).",
+)
+
+flags.DEFINE_integer(
+    "batch_size",
+    5,
+    "Number of circuits to send over the wire per batch (value does not affect results).",
+)
+
+flags.DEFINE_integer(
+    "n_shots", 2000, "Number of measurements to draw from each individual circuit."
+)
+
+flags.DEFINE_string(
+    "save_dir", "./data", "Path to save experiment data (must already exist)."
+)
+
+flags.DEFINE_bool("use_engine", False, "Whether or not to use quantum engine.")
+
+
+def flatten_circuit(circuit):
+    """Pack operations in circuit to the left as far as possible."""
+    return cirq.Circuit([op for mom in circuit for op in mom])
+
+
+def make_full_circuit(qubit_pairs, tsym, depth):
+    """Generate 1D tsym or scrambling circuit with given depth on qubit_pairs."""
+    inter_gen = None
+    if tsym == 1:
+        # Use tsym gates.
+        inter_gen = circuit_blocks.tsym_block
+    elif tsym == 0:
+        # use scrambling gates.
+        inter_gen = circuit_blocks.scrambling_block
+
+    # Random source for circuits.
+    random_source = np.random.uniform(0, 4, size=(depth * len(qubit_pairs), 2))
+
+    ret_circuit = cirq.Circuit()
+    tmp = circuit_blocks.block_1d_circuit(
+        [qubit[0] for qubit in qubit_pairs], depth, inter_gen, random_source
+    )
+    ret_circuit += flatten_circuit(tmp)
+
+    ret_circuit += cirq.Circuit(cirq.S(qubit[0]) for qubit in qubit_pairs)
+    ret_circuit += cirq.Circuit(cirq.H(qubit[0]) for qubit in qubit_pairs)
+
+    # Merge single qubit gates together and add measurements.
+    cirq.optimizers.merge_single_qubit_gates_into_phxz(ret_circuit)
+    cirq.DropEmptyMoments().optimize_circuit(circuit=ret_circuit)
+    ret_circuit = flatten_circuit(ret_circuit)
+
+    for i, qubit in enumerate([sublist[0] for sublist in qubit_pairs]):
+        ret_circuit += cirq.measure(qubit, key="q{}".format(i))
+
+    cirq.SynchronizeTerminalMeasurements().optimize_circuit(circuit=ret_circuit)
+    logging.debug(
+        f"Generated a new circuit w/ tsym={tsym} and depth {len(ret_circuit)}"
+    )
+    return ret_circuit
+
+
+def _engine_sim_workaround(batch):
+    """Use either simulator or engine."""
+    if FLAGS.use_engine:
+        project_id = os.environ["GOOGLE_CLOUD_PROJECT"]  # "quantum-simulation"
+        engine = cirq_google.Engine(project_id=project_id)
+        job = engine.run_batch(
+            programs=batch,
+            gate_set=cirq_google.SYC_GATESET,
+            processor_ids=["weber"],
+            repetitions=FLAGS.n_shots,
+        )
+        return job.results()
+
+    # Workaround for simulator.run_batch which doesn't have consistent return
+    # types with engine.run_batch.
+    sim = cirq.Simulator()
+    return [sim.run(circuit, repetitions=FLAGS.n_shots) for circuit in batch]
```

Comment:
open and link a cirq issue for this

Reply:
I'm not sure if this is a Cirq issue or my misuse of the library. I removed the comment and managed to convince myself it was my own misunderstanding.